src/transports/smart_protocol.c

```diff
@@ -270,50 +270,6 @@ static int store_common(transport_smart *t)
 	return 0;
 }
 
-static int fetch_setup_walk(git_revwalk **out, git_repository *repo)
-{
-	git_revwalk *walk = NULL;
-	git_strarray refs;
-	unsigned int i;
-	git_reference *ref = NULL;
-	int error;
-
-	if ((error = git_reference_list(&refs, repo)) < 0)
-		return error;
-
-	if ((error = git_revwalk_new(&walk, repo)) < 0)
-		return error;
-
-	git_revwalk_sorting(walk, GIT_SORT_TIME);
```

Comment:
By the way, one more question: previously, we've been sorting by time, now we aren't. While obvious that this will greatly improve performance as we now don't have to go through all commits, it's not immediately obvious to me why this is fine and didn't find anything in the commit message.

Reply:
True, I took @carlosmn's comment about blanket-replacing at face-value 😉. Checking around, [pack-protocol](https://github.com/git/git/blob/master/Documentation/technical/pack-protocol.txt#L268) seems to imply there's no expected order so this seems fine — though a second opinion would be welcome.

Reply:
Diving into the code a bit more. Previously, we've sorted all commits by date, which would cause us to compare the newest 256 commits present in our own repos with the HAVEs of the remote. Now with `push_glob`, we're pusing the commits in an unspecified order that is completely independent of the date. If one has a repo with more than 256 refs, then it may happen that we now compare 256 _old_ commits with the HAVEs. Heuristically, it's obvious that the first variant which adds commits by date may be a lot more efficient in the common case.

This analysis neglects that previously we had to load all commits of the repo to properly sort them by time, which was quite inefficient, too. I think we can do better than that, though: if `push_glob` were to use `git_commit_list_insert_by_date`, then we'd avoid inserting commits in a suboptimal order but we still wouldn't suck in all the repo's commit objects.

So I'd propose to modify `push_glob` to accept another parameter that determines how to sort pushed commits. `push_glob` will pass than on to `push_commit`, which will either use `git_commit_list_insert` or `git_commit_list_insert_by_date`.

Reply:
Thanks for the in-depth analysis @pks-t. I've amended the PR by API-fying a bit more (DRYing all the functions), so that the revwalk could be tweaked accordingly. Not sure if it's worth making this API public instead of the current one — no idea how useful it would be to "outsiders" — so it's a little rough around the edges (a single `flags` might have been sufficient).

Reply:
(though re-reading I'm now unsure if the insertion was to happen from an "initially sorted revwalk", as previously, or if it's okay with just the insertion being done by commit date. Right now the latter is done).